File: files/300-399/383/CH-383.R
library(tidyverse)
library(readxl)

path <- "300-399/383/CH-383 Filter.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "F3:F5")

is_valid <- function(s) {
  digits <- str_extract_all(s, "\\d")[[1]] %>% as.integer()
  digits[1] %% 2 == 0 && digits[length(digits)] %% 2 == 1
}

result <- input %>%
  filter(map_lgl(ID, is_valid))

all.equal(result$ID, test$ID)
# [1] TRUE

File: files/300-399/383/CH-383.py
import pandas as pd

path = "300-399/383/CH-383 Filter.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=2, nrows=8)
test = pd.read_excel(path, usecols="F", skiprows=2, nrows=2)

def is_valid(s):
    digits = [c for c in str(s) if c.isdigit()]
    return int(digits[0]) % 2 == 0 and int(digits[-1]) % 2 == 1

result = input[input["ID"].apply(is_valid)].reset_index(drop=True)

print(result["ID"].equals(test["ID.1"]))
# True
